File: src/mk3/JavaPackage/src/main/java/tamaf/ACLMessageTemplate.java
package tamaf;

import java.util.HashSet;
import java.util.Set;

public class ACLMessageTemplate {
    private AgentID sender;
    private Set<Performative> performatives;
    private String conversationId;
    private String protocol;
    private String ontology;

    public ACLMessageTemplate() {
        this.performatives = new HashSet<>();
    }

    public ACLMessageTemplate(AgentID sender, Set<Performative> performatives, String conversationId, String protocol, String ontology) {
        this.sender = sender;

        if (performatives == null) {
            this.performatives = new HashSet<>();
        } else {
            this.performatives = performatives;
        }

        this.conversationId = conversationId;
        this.protocol = protocol;
        this.ontology = ontology;
    }

    public boolean matches(ACLMessage msg) {
        if (sender != null) {
            if (msg.getSender() == null || !AgentID.matches(sender, msg.getSender())) {
                return false;
            }
        }
        
        if (performatives != null && !performatives.isEmpty()) {
            if (!performatives.contains(msg.getPerformative())) {
                return false;
            }
        }
        
        if (conversationId != null && !conversationId.equals(msg.getConversationId())) {
            return false;
        }
        
        if (protocol != null && !protocol.equals(msg.getProtocol())) {
            return false;
        }
        
        if (ontology != null && !ontology.equals(msg.getOntology())) {
            return false;
        }

        return true;
    }

    // Getters and Setters
    public AgentID getSender() { return sender; }
    public void setSender(AgentID sender) { this.sender = sender; }
    public Set<Performative> getPerformatives() { return performatives; }
    public void setPerformatives(Set<Performative> performatives) { this.performatives = performatives; }
    public void addPerformative(Performative performative) { this.performatives.add(performative); }
    public String getConversationId() { return conversationId; }
    public void setConversationId(String conversationId) { this.conversationId = conversationId; }
    public String getProtocol() { return protocol; }
    public void setProtocol(String protocol) { this.protocol = protocol; }
    public String getOntology() { return ontology; }
    public void setOntology(String ontology) { this.ontology = ontology; }

    // Static factory helpers
    public static ACLMessageTemplate MatchPerformative(Performative performative) {
        ACLMessageTemplate template = new ACLMessageTemplate();
        template.addPerformative(performative);
        return template;
    }

    public static ACLMessageTemplate MatchConversationId(String conversationId) {
        ACLMessageTemplate template = new ACLMessageTemplate();
        template.setConversationId(conversationId);
        return template;
    }

    public static ACLMessageTemplate MatchSender(AgentID sender) {
        ACLMessageTemplate template = new ACLMessageTemplate();
        template.setSender(sender);
        return template;
    }
}

File: src/mk3/PythonPackage/src/tamaf/aclmessagetemplate.py
from typing import TYPE_CHECKING
from .performative import Performative

if TYPE_CHECKING:
    from .core import AgentID
    from .aclmessage import ACLMessage

class ACLMessageTemplate:
    
    def __init__(self, sender: 'AgentID' = None, performatives: set[Performative] = None, conversationid: str = None, protocol: str = None, ontology: str = None, conversationID: str = None):
        self.sender = sender
        self.performatives = performatives
        self.conversationid = conversationid if conversationid is not None else conversationID
        self.protocol = protocol
        self.ontology = ontology

    @property
    def conversationID(self):
        return self.conversationid

    @conversationID.setter
    def conversationID(self, value):
        self.conversationid = value

    def SetPerformative(self, performative: Performative):

        if self.performatives is None:
            self.performatives = set()

        self.performatives.add(performative)

    def SetOntology(self, ontology: str):
        self.ontology = ontology

    def Matches(self, message: 'ACLMessage') -> bool:
        from .agentid import AgentID

        if self.sender is not None:
            if not AgentID.Matches(self.sender, message.sender):
                return False
        
        if self.performatives is not None:
            if message.performative not in self.performatives:
                return False
        
        if self.conversationid is not None:
            if message.conversationid != self.conversationid:
                return False
        
        if self.protocol is not None:
            if message.protocol != self.protocol:
                return False
        
        if self.ontology is not None:
            if message.ontology != self.ontology:
                return False
                
        return True

    @classmethod
    def MatchPerformative(cls, performative: Performative) -> 'ACLMessageTemplate':
        tmpl = cls()
        tmpl.SetPerformative(performative)
        return tmpl

    @classmethod
    def MatchConversationId(cls, conversationid: str) -> 'ACLMessageTemplate':
        return cls(conversationid=conversationid)

    @classmethod
    def MatchSender(cls, sender: 'AgentID') -> 'ACLMessageTemplate':
        return cls(sender=sender)
